### PR TITLE
Improve performance while getting dependencies of deployment JARs

### LINKED ISSUE #207: Improve performance while getting dependencies of deployment JARs
Collectio, of dependencies should be fast, but when dependencies are get for deployment JARs [here](https://github.com/redhat-developer/quarkus-ls/blob/7c555b9351c1ea743c4368702dd9864695171cfa/microprofile.jdt/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/providers/QuarkusConfigRootProvider.java#L144), it take so many time.

We should improve that by caching dependencies or perhaps get the dependencies with a kind of batch.

### PR DESCRIPTION
**Pull request by angelozerr**
Improve performance while getting dependencies of deployment JARs

Fixes #207

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR remove the full collection of dependencies for deployment JAR. The collection of dependencies is done just if it is required. 

To manage that, I did the same thing than scanning of classes, annotations to collect properties, in otherwise I manage a BuildingScopeContext which is a context used to build new classpath (like for Quarkus add deployment JAR). 

Each property providers can access to this context. I create a QuarkusContext which is shared between any properties provider when building scope is done. The QuarkusContext  gives the capability to add deployment artifact id for searching dependencies.

For instance for `quarkus-kubernetes-deployment` artifact, we need to search dependencies to add it in the search classpath (dekorate kubernetes class). This is done in QuarkusKubernetesPropertiesProvider with

```java
public void beginBuildingScope(BuildingScopeContext context, IProgressMonitor monitor) {
	QuarkusContext quarkusContext = QuarkusContext.getQuarkusContext(context);
	quarkusContext.getDependenciesToCollect().add("quarkus-kubernetes-deployment");
}

```

See https://github.com/redhat-developer/quarkus-ls/pull/222/files#diff-cb566d1c0b7847f1c453ae75f584bdacR81

Here a an array which compares time to collect properties for 2 projects:

| Project                | From master | From this PR |
|------------------------|-------------|--------------|
| all-quarkus-extension  | 33134ms     | 26689ms      |
| hibernate-orm-resteasy | 2815ms      | 1666ms       |

As you can see we win around 6 seconds for a big project like all-quarkus-extension and around 1 second for little project.

**Comment by angelozerr** [comment]
@xorye could you just test this PR and tell me if you have the feeling that it's less slow.

For the code, please wait I clean my code.

**Comment by xorye** [comment]
Very cool! I am getting significant performance improvements on my machine.

I noticed that I am missing this property:
```
  {
      "type": "java.lang.String",
      "sourceField": "path",
      "defaultValue": "/openapi",
      "extensionName": "quarkus-smallrye-openapi-common",
      "required": false,
      "phase": 1,
      "name": "quarkus.smallrye-openapi.path",
      "description": "The path at which to register the OpenAPI Servlet.",
      "sourceType": "io.quarkus.smallrye.openapi.common.deployment.SmallRyeOpenApiConfig"
    },
```

The `quarkus.smallrye-openapi.path` property is available on master, however.

After comparing all properties from the all-quarkus-extensions test project with master and this PR branch, that was the only missing property.

File: microprofile.jdt/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/providers/QuarkusKubernetesProvider.java [hunk 1]
```diff
@@ -74,6 +75,23 @@
...
+		// declared in the pom.xml
+		// When quarkus-kubernetes is declared, this JAR declares the deployment JAR
+		// quarkus-kubernetes-deployment
+		// This quarkus-kubernetes-deployment artifact has some dependencies to
+		// io.dekorate
+
+		// In other words, to add
+		// io.dekorate.kubernetes.annotation.KubernetesApplication class in the search
+		// classpath,
+		// the dependencies of quarkus-kubernetes-deployment artifact must be done
```
Comment (xorye) [inline microprofile.jdt/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/providers/QuarkusKubernetesProvider.java]: Should this say "must be downloaded"?

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/internal/core/FakeJavaProject.java [hunk 2]
```diff
@@ -58,13 +61,15 @@ public FakeJavaProject(IJavaProject rootProject, IClasspathEntry[] entries) thro
...
 		}
+		List<IJavaElement> elements = new ArrayList<>();
+		elements.add(rootProject);
 		IPackageFragmentRoot[] roots = super.getPackageFragmentRoots();
-		IJavaElement[] elements = new IJavaElement[1 + roots.length];
-		elements[0] = rootProject;
-		for (int i = 0; i < roots.length; i++) {
-			elements[i + 1] = roots[i];
+		for (IPackageFragmentRoot root : roots) {
+			if (searchClasspathEntries.contains(root.getResolvedClasspathEntry())) {
```
Comment (xorye) [inline microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/internal/core/FakeJavaProject.java]: Could you please explain what this check is for? I'm having trouble understanding this statement

Comment (xorye) [reply microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/internal/core/FakeJavaProject.java]: Filtering classpath entries to get the ones we need? It looks like `QuarkusConfigRootProvider.contributeToClasspath()` adds entries to `searchClasspathEntries`. 

Should `searchClasspathEntries` only contain deployment artifacts?

Comment (angelozerr) [reply microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/internal/core/FakeJavaProject.java]: I remove this code. Don't need to do that.

**Comment by angelozerr** [comment]
> The quarkus.smallrye-openapi.path property is available on master, however.

Good catch, it should work now.

**Comment by xorye** [comment]
Awesome, the PR works great

File: microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/PropertiesManager.java [hunk 3]
```diff
@@ -289,24 +288,36 @@ private IJavaProject getJavaProject(IJavaProject javaProject, boolean excludeTes
...
+				ArtifactResolver.DEFAULT_ARTIFACT_RESOLVER);
+		beginBuildingScope(context, mainMonitor);
 		for (int i = 0; i < length; i++) {
 			mainMonitor.subTask("Contributing to classpath for provider (" + (i + 1) + "/" + length + ")");
 			SubMonitor subMonitor = mainMonitor.split(1);
 			IPropertiesProvider provider = getPropertiesProviders().get(i);
-			provider.contributeToClasspath(javaProject, resolvedClasspath, excludeTestCode, scopes,
-					ArtifactResolver.DEFAULT_ARTIFACT_RESOLVER, newClasspathEntries, subMonitor);
+			provider.contributeToClasspath(context, subMonitor);
 			subMonitor.done();
```
Comment (xorye) [inline microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/PropertiesManager.java]: Could we have this in another method like `beginBuildingScope()` and `endBuildingScope()`, just to clean things up and be consistent?

Like:
```
private void contributeToClasspath(BuildingScopeContext context, IProgressMonitor monitor) throws JavaModelException {
	int length = getPropertiesProviders().size();
	for (int i = 0; i < length; i++) {
		monitor.subTask("Contributing to classpath for provider (" + (i + 1) + "/" + length + ")");
		SubMonitor subMonitor = ((SubMonitor)monitor).split(1);
		IPropertiesProvider provider = getPropertiesProviders().get(i);
		provider.contributeToClasspath(context, subMonitor);
		subMonitor.done();
	}
}
```

Comment (angelozerr) [reply microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/PropertiesManager.java]: Why do you want to do that? The main idea (like collection of properties) is to have a begin (o initialize something), process (do some work which can take time) and end (dispose something) methods.

Comment (angelozerr) [reply microprofile.jdt/com.redhat.microprofile.jdt.core/src/main/java/com/redhat/microprofile/jdt/core/PropertiesManager.java]: @xorye updated

**Pull request merged**